deps/undici/src/lib/observable.js

```diff
@@ -0,0 +1,802 @@
+                  } else {
+                    while (deferred.length > 0) {
+                      const [, reject] = deferred.shift();
+                      reject(err);
+                    }
+                  }
+                },
+                complete() {},
+              },
+              {
+                signal: ac.signal,
+              }
+            );
+          }
+          deferred.push([resolve, reject]);
+        });
+      },
+      throw: (err) => {
+        return new Promise((_resolve, reject) => {
+          ac?.abort();
+          hasError = true;
+          error = err;
+          for (const [, deferredReject] of deferred) {
+            deferredReject(error);
+          }
+          reject(error);
+        });
+      },
+      return: () => {
+        return new Promise((resolve) => {
+          ac?.abort();
+          isComplete = true;
+          for (const [deferredResolve] of deferred) {
+            deferredResolve({ value: undefined, done: true });
+          }
+          resolve({ value: undefined, done: true });
+        });
+      },
+      [Symbol.asyncIterator]() {
+        return this;
+      },
+    };
+  }
+}
+
+class Subscriber {
+  #active = true;
+  #abortController = new AbortController();
+  #signal;
+  #teardowns;
+
+  get signal() {
+    return this.#signal;
+  }
+
+  get isActive() {
+    return this.#active && !this.signal.aborted;
+  }
+
+  constructor(signal, _observer) {
+    this._observer = _observer;
+    const ownSignal = this.#abortController.signal;
+    this.#signal = signal ? AbortSignal.all([signal, ownSignal]) : ownSignal;
+  }
+
+  next(value) {
+    if (this.isActive) {
+      this._observer.next(value);
+    }
+  }
+
+  error(error) {
+    if (this.isActive) {
+      this.#active = false;
+      this._observer.error(error);
+      this.#abortController.abort();
+    }
+  }
+
+  complete() {
+    if (this.isActive) {
+      this.#active = false;
+      this._observer.complete();
+      this.#abortController.abort();
+    }
+  }
+
+  #teardownHandler = () => {
+    const teardowns = Array.from(this.#teardowns);
+    this.#teardowns = undefined;
+    let errors;
+    for (let i = teardowns.length - 1; i >= 0; i--) {
+      try {
+        teardowns[i]();
+      } catch (error) {
+        errors ??= [];
+        errors.push(error);
+      }
+    }
+    if (errors) {
+      // TODO: We need to figure out how to report multiple errors.
+      throw new Error(
+        `${errors.length} teardowns failed.\n ${errors.join('\n')}}`
+      );
+    }
+  };
+
+  addTeardown(teardown) {
+    if (this.isActive) {
+      if (!this.#teardowns) {
+        this.#teardowns = [];
+        this.#abortController.signal.addEventListener(
+          'abort',
+          this.#teardownHandler,
+          {
+            once: true,
+          }
+        );
+      }
+      this.#teardowns.push(teardown);
+    } else {
+      teardown();
+    }
+  }
+
+  removeTeardown(teardown) {
```

Comment:
I don't _think_ this is in the proposal yet, but it should be. There are edge cases where teardowns should be removed.

Reply:
This probably shouldn't appear in the reference implementation then.